ARCH: introduce functionality for forcing core architecture HW initialization during boot, for chain-loadable images.

ioannisg opened the pull request:
This is still draft for now
- Goal: introduce functionality for forcing core architecture HW initialization during boot, for chain-loadable images
- Implemented for Cortex-M
- Used when building Zephyr with MCUboot

Fixes #7404
Addresses #28803 for ARM cortex-M
Relates to the MCUboot tickets
- JuulLabs-OSS/mcuboot#833
- JuulLabs-OSS/mcuboot#824 
- JuulLabs-OSS/mcuboot#828

ioannisg committed:
arch: introduce option to force internal architectural state init

We introduce an option that instructs Zephyr to perform
the initialization of internal architectural state (e.g.
ARCH-level HW registers and system control blocks) during
early boot to the reset values. The option is available
to the application developer but shall depend on whether
the architecture supports the functionality.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: cortex_m: implement functionality for ARCH core regs init

Implement the functionality for configuring the
architecture core registers to their warm reset
values upon system initialization. We enable the
support of the feature in the Cortex-M architecture.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
zephyr: make images loaded by MCUboot enable arch core HW regs init

Instruct Zephyr images that are loaded by MCUboot to
force the initialization of architecture core HW registers,
if the architecture supports the functionality.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

nashif commented:
please put something meaningful in the PR title, not just the default branch name

ioannisg commented:
@abrodkin could you take a look at this? I'd like this feature to be cross-ARCH, and supported, initially at least, on ARC and ARM that seem to have issues with initialization of ARCH-core hardware in certain use cases. So any initial feedback on Kconfig option and text is welcome.

ioannisg commented on arch/Kconfig:
```diff
@@ -381,6 +381,22 @@ config EXTRA_EXCEPTION_INFO
+	  values. The option is useful when the Zephyr firmware
+	  image is chain-loaded by e.g. a bootloader and we need
+	  to guarantee that the architecture core HW registers
+	  are initialized to their warm reset values (as specified
```
Actually no, @nvlsianpu (but this is a good question). Warm reset conditions is the minimum we need to satisfy, in general, so no need to set cold reset here. 

Might as well remove "warm" and leave just "reset" if this seems confusing :)

nvlsianpu replied:
Better to clarify what this kind of reset values means - having adjective her is good as this is certain kind of reset values

abrodkin replied:
IMHO description is a bit incomplete or even incorrect. Registers are just an interface to communicate with hardware. That said instead of setting some registers in its pristine state we want to get some HW internals back to its non-adultered state.

And an example is ARConnect IP (set of IP blocks to be used on SMP systems). There we have an interface of just 3 registers (one for command, one for data to write, one for data to read) for dealing with many very different blocks like cross-core IRQs, independent timer etc. And setting those 3 regs with just 1 set of values won't do the work.

ioannisg replied:
@abrodkin thanks
So you propose to reword "HW registers" to "Architecture core HW internals"? or something similar?

abrodkin replied:
Yep, to something like "Initialize internal architectural states at boot".

ioannisg replied:
Yeap, that's fine with me. Updated now @abrodkin.

abrodkin commented on arch/Kconfig:
```diff
@@ -381,6 +381,22 @@ config EXTRA_EXCEPTION_INFO
+	  Note: the functionality is architecture-specific. For the
+	  implementation details refer to each architecture(s) where
+	  the feature is supported.
+
```
I think it's not only something architecture specific but at least as well might be SoC-related and thus expand over purely architecture code.

Or do we plan to have yet another similar option for SoC's and boards?

ioannisg replied:
Yes, for now this PR concerns only what can be defined as common ARCH-level HW registers. SoCs and Boards have the PLATFORM_INIT Kconfig to set (or reset) specific hardware. So it is not in the scope of this PR.

ioannisg commented:
Adding @andrewboie to get feedback on the added Kconfig option at the arch-level

abrodkin commented:
@ioannisg thanks! Though corresponding commit title and message should be updated as well?
```
arch: introduce option to force architecture core registers init

We introduce an option that instructs Zephyr to perform
an initialization of architecture core HW registers upon
boot. The options is user configurable but shall depend
on whether the architecture supports the functionality.
```

ioannisg commented:
> @ioannisg thanks! Though corresponding commit title and message should be updated as well?
> 
> ```
> arch: introduce option to force architecture core registers init
> 
> We introduce an option that instructs Zephyr to perform
> an initialization of architecture core HW registers upon
> boot. The options is user configurable but shall depend
> on whether the architecture supports the functionality.
> ```

Agreed @abrodkin , thanks for catching this.
Done now.

Btw, any chance you can use this for ARC?

abrodkin commented:
You mean re-work https://github.com/zephyrproject-rtos/zephyr/pull/28494?

ioannisg commented:
> You mean re-work #28494?

Yes, that's what I had in mind. But you might want to do more under CONFIG_INIT_ARCH_HW_AT_BOOT, it's your choice, @abrodkin.

That should not affect the review of this PR. I was just curious.

abrodkin commented:
@YuguoWH could you please switch ARC on `CONFIG_INIT_ARCH_HW_AT_BOOT` instead of our very own `CONFIG_ARC_CUSTOM_INIT`?

nvlsianpu commented:
@ioannisg Can you add record about this feature to the release-note?

nvlsianpu reviewed (changes requested):
I have got bootloader jumping to application and reset looping while using with helloWord sample.
For SMP_SVR application hangs.

ioannisg committed:
doc: mention the INIT_ARCH_HW_AT_BOOT functionality in the rel notes

Mention the newly introduced funtionality for Cortex-M
architecture to force the early-boot initialization of
internal architecture HW state.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
> @ioannisg Can you add record about this feature to the release-note?

Yeap, I added a Release-Notes label and pushed a release-notes commit.

ioannisg committed:
arch: arm: aarch32: cortex_m: improve documentation of z_arm_reset

We enhance the documentation of z_arm_reset, stressing that
the function may either be loaded by the processor coming
out of reset, or by another image, e.g. a bootloader. We
also specify what is required at minimum when executing the
reset function.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

nvlsianpu commented on west.yml:
```diff
@@ -93,8 +93,9 @@ manifest:
       path: modules/crypto/mbedtls
     - name: mcuboot
-      revision: e64c5f00d67cf858d1fdead7135f5d61de2a4c68
+      revision: pull/847/head
```
PR for mcuboot synch https://github.com/zephyrproject-rtos/mcuboot/pull/38

ioannisg replied:
Noticed that, thanks. Just rebased the PR and the manifest update commit went away :+1:

ioannisg commented:
Removed DNM flag since manifest now points to the mcuboot SHA containing the corresponding updates in the mcuboot side.

ioannisg commented:
@carlescufi pls, take a look at the commit introducing the  cross-arch Kconfig option.

carlescufi commented on Kconfig.zephyr:
```diff
@@ -397,6 +398,11 @@ config BOOTLOADER_MCUBOOT
 
+	  By default, this option instructs Zephyr to initialize the core
+	  architecture HW registers during boot, when this is supported by
+	  the application. This removes the need by MCUboot to reset itself
+	  the core registers' state.
```
```suggestion
	  the application. This removes the need by MCUboot to reset
	  the core registers' state itself.
```

Prevent confusion about what is reset.

carlescufi commented on arch/Kconfig:
```diff
@@ -384,6 +384,23 @@ config EXTRA_EXCEPTION_INFO
+	depends on ARCH_SUPPORTS_ARCH_HW_INIT
+	help
+	  This option instructs Zephyr to force the initialization
+	  of internal architectural state (for example ARCH-level
```
```suggestion
	  of the internal architectural state (for example ARCH-level
```

carlescufi commented on arch/Kconfig:
```diff
@@ -384,6 +384,23 @@ config EXTRA_EXCEPTION_INFO
+	  This option instructs Zephyr to force the initialization
+	  of internal architectural state (for example ARCH-level
+	  HW registers and system control blocks) during boot to
+	  the reset values. The option is useful when the Zephyr
```
```suggestion
	  the reset values as specified by the corresponding architecture manual. The option is useful when the Zephyr
```

ioannisg replied:
done

carlescufi reviewed (approved):
Just a couple of minor comments. Other than that, I am happy with the introduction of this new option.
